File: packages/rslint-test-tools/tests/typescript-eslint/rules/no-useless-empty-export.test.ts
import { describe, test, expect } from '@rstest/core';
import { noFormat, RuleTester } from '@typescript-eslint/rule-tester';
import { getFixturesRootDir } from '../RuleTester.ts';

const rootPath = getFixturesRootDir();
// this rule tests the spacing, which prettier will want to fix and break the tests
/* eslint "@typescript-eslint/internal/plugin-test-formatting": ["error", { formatWithPrettier: false }] */

const ruleTester = new RuleTester({
  languageOptions: {
    parserOptions: {
      project: './tsconfig.json',
      tsconfigRootDir: rootPath,
    },
  },
});

const error = {
  messageId: 'uselessExport',
} as const;

describe('no-useless-empty-export', () => {
  test('should work', () => {
    ruleTester.run('no-useless-empty-export', {
      valid: [
        "declare module '_'",
        "import {} from '_';",
        "import * as _ from '_';",
        'export = {};',
        'export = 3;',
        'export const _ = {};',
        `
      const _ = {};
      export default _;
    `,
        `
      export * from '_';
      export = {};
    `,
        `
      export {};
    `,
        // https://github.com/microsoft/TypeScript/issues/38592
        {
          code: `
        export type A = 1;
        export {};
      `,
          // @ts-ignore
          filename: 'foo.d.ts',
        },
        {
          code: `
        export declare const a = 2;
        export {};
      `,
          // @ts-ignore
          filename: 'foo.d.ts',
        },
        {
          code: `
        import type { A } from '_';
        export {};
      `,
          // @ts-ignore
          filename: 'foo.d.ts',
        },
        {
          code: `
        import { A } from '_';
        export {};
      `,
          // @ts-ignore
          filename: 'foo.d.ts',
        },
      ],
      invalid: [
        {
          code: `
export const _ = {};
export {};
      `,
          errors: [error],
          output: `
export const _ = {};

      `,
        },
        {
          code: `
export * from '_';
export {};
      `,
          errors: [error],
          output: `
export * from '_';

      `,
        },
        {
          code: `
export {};
export * from '_';
      `,
          errors: [error],
          output: `

export * from '_';
      `,
        },
        {
          code: `
const _ = {};
export default _;
export {};
      `,
          errors: [error],
          output: `
const _ = {};
export default _;

      `,
        },
        {
          code: `
export {};
const _ = {};
export default _;
      `,
          errors: [error],
          output: `

const _ = {};
export default _;
      `,
        },
        {
          code: `
const _ = {};
export { _ };
export {};
      `,
          errors: [error],
          output: `
const _ = {};
export { _ };

      `,
        },
        {
          code: `
import _ = require('_');
export {};
      `,
          errors: [error],
          output: `
import _ = require('_');

      `,
        },
        {
          code: `
import _ = require('_');
export {};
export {};
      `,
          errors: [error, error],
          output: `
import _ = require('_');


      `,
        },
      ],
    });
  });
});

File: internal/plugins/typescript/rules/no_useless_empty_export/no_useless_empty_export_test.go
package no_useless_empty_export

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/typescript/rules/fixtures"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestNoUselessEmptyExportRule(t *testing.T) {
	rule_tester.RunRuleTester(fixtures.GetRootDir(), "tsconfig.json", t, &NoUselessEmptyExportRule, []rule_tester.ValidTestCase{
		{Code: "declare module '_'"},
		{Code: "import {} from '_';"},
		{Code: "import * as _ from '_';"},
		{Code: "export = {};"},
		{Code: "export = 3;"},
		{Code: "export const _ = {};"},
		{Code: `
const _ = {};
export default _;
`},
		{Code: `
export * from '_';
export = {};
`},
		{Code: `
export {};
`},
		// https://github.com/microsoft/TypeScript/issues/38592
		{
			Code: `
export type A = 1;
export {};
`,
		},
		{
			Code: `
export declare const a = 2;
export {};
`,
		},
		{
			Code: `
import type { A } from '_';
export {};
`,
		},
	}, []rule_tester.InvalidTestCase{
		{
			Code: `
export const _ = {};
export {};
`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "uselessExport",
					Line:      3,
					Column:    1,
				},
			},
			Output: []string{`
export const _ = {};

`},
		},
		{
			Code: `
export * from '_';
export {};
`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "uselessExport",
					Line:      3,
					Column:    1,
				},
			},
			Output: []string{`
export * from '_';

`},
		},
		{
			Code: `
export {};
export * from '_';
`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "uselessExport",
					Line:      2,
					Column:    1,
				},
			},
			Output: []string{`

export * from '_';
`},
		},
		{
			Code: `
const _ = {};
export default _;
export {};
`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "uselessExport",
					Line:      4,
					Column:    1,
				},
			},
			Output: []string{`
const _ = {};
export default _;

`},
		},
		{
			Code: `
export {};
const _ = {};
export default _;
`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "uselessExport",
					Line:      2,
					Column:    1,
				},
			},
			Output: []string{`

const _ = {};
export default _;
`},
		},
		{
			Code: `
const _ = {};
export { _ };
export {};
`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "uselessExport",
					Line:      4,
					Column:    1,
				},
			},
			Output: []string{`
const _ = {};
export { _ };

`},
		},
		{
			Code: `
import _ = require('_');
export {};
`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "uselessExport",
					Line:      3,
					Column:    1,
				},
			},
			Output: []string{`
import _ = require('_');

`},
		},
		{
			Code: `
import _ = require('_');
export {};
export {};
`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "uselessExport",
					Line:      3,
					Column:    1,
				},
				{
					MessageId: "uselessExport",
					Line:      4,
					Column:    1,
				},
			},
			Output: []string{`
import _ = require('_');


`},
		},
		{
			Code: `
import { A } from '_';
export {};
`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "uselessExport",
					Line:      3,
					Column:    1,
				},
			},
			Output: []string{`
import { A } from '_';

`},
		},
	})
}
